bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java

```diff
@@ -0,0 +1,265 @@
+                            previousLocation, currentLocation);
+                    createResults(currentLocation);
+                    previousLocation = currentLocation;
+                }
+            }, 0, LOCATION_CHANGED_CHECK_INTERVAL_SECONDS, TimeUnit.SECONDS);
+            logger.debug("Scheduled FMI Weather location-changed discovery job every {} seconds",
+                    LOCATION_CHANGED_CHECK_INTERVAL_SECONDS);
+        }
+    }
+
+    public void createResults(@Nullable PointType location) {
+        createForecastForCurrentLocation(location);
+        createForecastsForCities(location);
+        createObservationsForStations(location);
+    }
+
+    private void createForecastForCurrentLocation(@Nullable PointType currentLocation) {
+        if (currentLocation != null) {
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(LOCATION,
+                    String.format("%s,%s", currentLocation.getLatitude(), currentLocation.getLongitude()));
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(UID_LOCAL_FORECAST)
+                    .withLabel(String.format("FMI local weather forecast")).withProperties(properties).build();
+            thingDiscovered(discoveryResult);
+        }
+    }
+
+    private void createForecastsForCities(@Nullable PointType currentLocation) {
+        List<Location> filteredCities = new LinkedList<>();
+        CITIES_OF_FINLAND.stream().filter(location2 -> isClose(currentLocation, location2)).peek(station -> {
+            if (logger.isDebugEnabled()) {
+                filteredCities.add(station);
+            }
+        }).forEach(city -> {
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(LOCATION,
+                    String.format("%s,%s", city.latitude.toPlainString(), city.longitude.toPlainString()));
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder
+                    .create(new ThingUID(THING_TYPE_FORECAST, cleanId(String.format("city_%s", city.name))))
+                    .withLabel(String.format("FMI weather forecast for %s", city.name)).withProperties(properties)
+                    .build();
+            thingDiscovered(discoveryResult);
+        });
+        if (logger.isDebugEnabled()) {
+            logger.debug("Filtered cities for forecasts: {}",
+                    filteredCities.stream().map(station -> String.format("%s (%s)", station.name, station.id))
+                            .collect(Collectors.toCollection(TreeSet<String>::new)));
+        }
+    }
+
+    private void createObservationsForStations(@Nullable PointType location) {
+        List<Location> candidateStations = new LinkedList<>();
+        List<Location> filteredStations = new LinkedList<>();
+        cachedStations().peek(station -> {
+            if (logger.isDebugEnabled()) {
```

Comment:
See above

Reply:
Keeping this one, I'm able to debug users discovery using logs only (spits out candidate stations and filtered/close-by stations)